ReSharper.FSharp/src/FSharp.Psi.Services/src/Generate/GenerateProvider.fs

```diff
@@ -198,28 +198,75 @@ type FSharpOverridingMembersBuilder() =
-        shiftNode (-currentIndent) typeRepr
-        shiftNode desiredIndent typeRepr
+        let normalizeReprEnd (beginToken: ITokenNode) (endToken: ITokenNode) =
+            if beginToken.Indent > endToken.Indent then
+                let diff = beginToken.Indent - endToken.Indent
+                addNodeBefore endToken (Whitespace(diff))
         
+        match typeRepr with
+        | :? IClassRepresentation as classRepr ->
+            normalizeReprEnd classRepr.BeginKeyword classRepr.EndKeyword
+        | :? IStructRepresentation as structRepr ->
+            normalizeReprEnd structRepr.BeginKeyword structRepr.EndKeyword
+        | :? IRecordRepresentation as recordRepr ->
+            normalizeReprEnd recordRepr.LeftBrace recordRepr.RightBrace
+
+            if recordRepr.FieldDeclarations.Count > 1 then
+                recordRepr.FieldDeclarations
+                |> Seq.indexed
+                |> Seq.skip 1
+                |> Seq.iter (fun (idx, f) ->
+                    let prevField = recordRepr.FieldDeclarations[idx - 1]
+                    if f.StartLine <> prevField.StartLine && f.Indent > recordRepr.FieldDeclarations[0].Indent then
+                        let diff = f.Indent - recordRepr.FieldDeclarations[0].Indent
+                        shiftWithWhitespaceBefore -diff f)
+        | _ -> ()
+
+        // members might need deindenting after the representation start was moved to it's own line
         typeDecl.TypeMembers
         |> Seq.iter (fun m ->
-            match m.GetPreviousToken() with
-            | :? Whitespace as whitespace ->
-                let diff = m.Indent - desiredIndent
-                if diff > 0 then
-                    shiftWhitespaceBefore -diff whitespace
-                    shiftNode -diff m
+            let diff = m.Indent - desiredIndent
+            if diff > 0 then shiftWithWhitespaceBefore -diff m)
+
+        let typeMembers =
+            match typeRepr with
+            | :? IObjectModelTypeRepresentation as repr ->
+                repr.TypeMembersEnumerable
+                |> Seq.filter (
+                    function
+                    | :? ILetBindingsDeclaration
+                    | :? IInterfaceImplementation -> true
+                    | _ -> false)
+                |> Seq.map (fun m -> m :> ITreeNode)
+            | _ -> List.empty
+        
+        let membersDecls = typeRepr.GetMemberDeclarations() |> Seq.map (fun m -> m :> ITreeNode)
+        let declarations = Seq.append typeMembers membersDecls
+
+        // members between representation start/end might need further indentation
+        declarations
+        |> Seq.iter (fun m ->
+            match m with
+            | :? IValFieldDeclaration
+            | :? IAbstractMemberDeclaration
+            | :? IConstructorDeclaration
+            | :? IRecordFieldDeclaration
+            | :? ILetBindingsDeclaration
+            | :? IInterfaceImplementation
+            | :? IMemberDeclaration ->
```

Comment:
It's been already discussed in the previous PRs, please try to use base interfaces instead of more specific ones where possible.

Reply:
Ah yeah, sorry. Got carried away there.